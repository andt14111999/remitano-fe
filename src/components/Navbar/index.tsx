import images from 'assets';
import CustomButton from 'components/CustomButton';
import CustomModal from 'components/CustomModal';
import React, { useState } from 'react';
import { createPortal } from 'react-dom';
import { useMediaQuery } from 'react-responsive';
import { Link } from 'react-router-dom';
import { useSpring, animated, useTransition, config } from 'react-spring';

const Navbar = () => {
  const [isMobileSidebarOpen, setIsMobileSidebarOpen] = useState(false);
  const animationSidebar = useSpring({
    to: {
      opacity: isMobileSidebarOpen ? 1 : 0,
      transform: isMobileSidebarOpen ? 'translateX(0%)' : 'translateX(100%)',
    },
  });

  const isDesktop = useMediaQuery({
    query: '(min-width: 1224px)',
  });
  const isTablet = useMediaQuery({
    minWidth: '768px',
    maxWidth: '1023px',
  });
  const isMobile = useMediaQuery({
    query: '(max-width: 767px)',
  });

  const handleOpenSidebar = () => {
    setIsMobileSidebarOpen(true);
  };

  const handleCloseSidebar = () => {
    setIsMobileSidebarOpen(false);
  };

  return (
    <>
      <div className="px-4 py-4 transition-all flex justify-between items-center font-medium border-b border-muted">
        <div className="flex items-center gap-12">
          <Link to="/">
            <img
              src={images.logo}
              alt="Remitano"
              className="w-[3.25rem] md:w-[4rem]"
            />
          </Link>
        </div>
        <div className="flex items-center gap-4 xl:gap-12">
          {isDesktop && (
            <Link to="/login">
              <CustomButton className="custom-button-shadow">
                Login / Register
              </CustomButton>
            </Link>
          )}
          {!isDesktop && (
            <img
              src="/images/Hamburger.svg"
              alt="Remitano"
              className="w-[2rem] h-[2rem]"
              onClick={() => {
                handleOpenSidebar();
              }}
            />
          )}
        </div>
      </div>
      {!isDesktop &&
        createPortal(
          <>
            <animated.div
              style={animationSidebar}
              className="fixed z-[100000] top-0 left-0 bottom-0 mx-0 px-0"
              // ref={ref}
            >
              <div className="h-full w-full">
                <div className="z-[100000] w-[100svw] h-[100svh] overflow-auto bg-neutral-100 flex flex-col">
                  <div className="px-6 py-5 flex justify-between border-b border-neutral-300">
                    <Link to="#">
                      <img
                        src="/images/Logo.svg"
                        alt="Remitano"
                        className="md:w-[4rem] h-[2.5rem] md:h-[3rem] 2xl"
                      />
                    </Link>
                    <img
                      src="/images/Navbar_Close.svg"
                      alt="Remitano"
                      className="md:w-[4rem] h-[2.5rem] md:h-[3rem] 2xl"
                      onClick={() => {
                        handleCloseSidebar();
                      }}
                    />
                  </div>

                  <div className="p-6">
                    <div className="mb-14 flex flex-col gap-6">
                      <a
                        href="#introduction"
                        className="text-neutral-black tracking-minus-01"
                        onClick={() => {
                          handleCloseSidebar();
                        }}
                      >
                        Giới thiệu chung
                      </a>
                      <a
                        href="#interaction"
                        className="text-neutral-black tracking-minus-01"
                        onClick={() => {
                          handleCloseSidebar();
                        }}
                      >
                        Chức năng
                      </a>
                      <a
                        href="#preview"
                        className="text-neutral-black tracking-minus-01"
                        onClick={() => {
                          handleCloseSidebar();
                        }}
                      >
                        Phụ kiện
                      </a>
                      <a
                        href="#dropdown"
                        className="text-neutral-black tracking-minus-01"
                        onClick={() => {
                          handleCloseSidebar();
                        }}
                      >
                        Hướng dẫn sử dụng
                      </a>
                      <a
                        href="#faq"
                        className="text-neutral-black tracking-minus-01"
                        onClick={() => {
                          handleCloseSidebar();
                        }}
                      >
                        FAQ
                      </a>
                    </div>
                  </div>
                </div>
              </div>
            </animated.div>
          </>,
          document.body
        )}
    </>
  );
};

export default Navbar;
