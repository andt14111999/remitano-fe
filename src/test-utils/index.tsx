import { renderHook } from '@testing-library/react';
import { MediaQueryMatchers, useMediaQuery } from 'react-responsive';

export const testMediaQuery = (query: MediaQueryMatchers) => {
  const { result } = renderHook(() => useMediaQuery(query));
  return result.current;
};

export const sleep = (ms: number) => {
  return new Promise((resolve) => setTimeout(resolve, ms));
};
