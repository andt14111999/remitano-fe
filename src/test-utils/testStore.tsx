import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import userSlice from 'stores/userSlice';
import uiSlice from 'stores/uiSlice';
import { AppStore, RootState, setupStore } from 'stores/store';
import { HelmetProvider } from 'react-helmet-async';
import { BrowserRouter } from 'react-router-dom';

// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    // store = setupStore({
    //   reducer: { ui: uiSlice.reducer, user: userSlice.reducer },
    //   preloadedState,
    // }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <HelmetProvider>
        <BrowserRouter>
          <Provider store={store}>{children}</Provider>
        </BrowserRouter>
      </HelmetProvider>
    );
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
