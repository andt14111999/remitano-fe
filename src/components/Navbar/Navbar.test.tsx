import { Context as ResponsiveContext } from 'react-responsive';
import Navbar from '.';
import { renderWithProviders } from 'test-utils/testStore';
import userEvent from '@testing-library/user-event';
import {
  logRoles,
  screen,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { sleep } from 'test-utils';
import { act } from 'react-dom/test-utils';

describe('Navbar', () => {
  test('mobile click on hamburger', async () => {
    const user = userEvent.setup();
    const { container, unmount } = renderWithProviders(
      <ResponsiveContext.Provider value={{ width: 300 }}>
        <Navbar />
      </ResponsiveContext.Provider>
    );
    logRoles(container);
    // click on hamburger
    const hamburger = await screen.findByRole('img', { name: /hamburger/i });
    await user.click(hamburger);

    // expect navbar to be visible
    const navbarClose = await screen.findByRole('img', { name: /close/i });
    expect(navbarClose).toBeInTheDocument();

    // click on navbar and check if the section have style
    const mobileWrapper = await screen.findByRole('menubar');

    await user.click(navbarClose);
    await sleep(1000);
    expect(mobileWrapper).toHaveStyle({
      opacity: 0,
      transform: 'translateX(100%)',
    });
    unmount();
  });
});
