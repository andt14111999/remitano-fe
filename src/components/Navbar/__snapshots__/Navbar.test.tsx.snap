// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsListing matches the snapshot 1`] = `
<div>
  <div
    class="px-4 py-4 transition-all flex justify-between items-center font-medium border-b border-muted"
  >
    <div
      class="flex items-center gap-12"
    >
      <a
        href="/"
      >
        <img
          alt="Remitano"
          class="w-[3.25rem] md:w-[4rem]"
          src="Collect.png"
        />
      </a>
    </div>
    <div
      class="flex items-center gap-4 xl:gap-12"
    >
      <a
        href="/login"
      >
        <button
          class="flex justify-center items-center gap-x-[0.625rem] p-2 bg-slate-100 rounded-lg px-4 py-2 border-2 border-fg-default text-fg-default shadow-lg custom-button-shadow "
          type="button"
        >
          <p
            class=" font-bold"
          >
            Login / Register
          </p>
        </button>
      </a>
    </div>
  </div>
</div>
`;
